### DISCUSSION TITLE
Importance of register_type()

### DISCUSSION THREAD

**Discussion by MuhammedErkam** [OpenDDS/OpenDDS#4083] [Q&A] [answered]
My question is I have 2 publisher and 1 subscriber. I gave my type and topic. But in one publisher I changed type name differently.

For example for pub1:

```
my_type = "dds nice";
my_topic = "dds topic"; 
```

For example for pub2:

```
my_type = "skjskjskjskj";
my_topic = "dds topic"; 
```

For example for sub:

```
my_type = "dds nice";
my_topic = "dds topic"; 
```

And then pub1 and pub2 works. I  didnt get it. Because in intro Page you said : 

> We first register the IDL Quote type with the domain participant, passing an instance of the generated QuoteTypeSupportImpl class for the Quote type. The name that we use for the Quote type, which is stored in the constant value QUOTER_QUOTE_TYPE, **MUST MATCH** the name used on the subscriber. When we create a topic, we specify this type name, enabling DCPS to later create the appropriate type of data writer for the topic.

Why it is working?

**Comment by iguessthislldo** [comment, marked as answer]
This appears to happen with RTPS, but not InfoRepo, so I'm assuming you're using RTPS. What you're seeing is part of the XTypes features of DDS that allows types in different participants to be different and still work together. In this case if OpenDDS can exchange type information then it can figure out if the types will work together and doesn't even check the type names. This comes from the XTypes specification version 1.3 section 7.6.3.4.2 "Rules for Type Consistency Enforcement":

> The type-consistency-enforcement rules consist of two steps.
Step 1. If both the Publication and the Subscription specify a TypeObject, consider it first.
...
Step 2. If either the Publication or the Subscription does not provide a TypeObject definition,
then the type names are consulted. The Subscription and Publication type_name fields must
match exactly, as in [DDS] prior to this specification.
...

I confirmed that's how this is working by modifying another example to work like what you're saying. I changed the publisher to use a different type name from the subscriber and it worked. However if I set [`UseXtypes=no`](https://opendds.readthedocs.io/en/master/devguide/run_time_configuration.html#run-time-configuration-usextypes) to stop it exchanging type details then it doesn't work anymore. It can't check step 1, so it goes to step 2 and compares the type names instead.

That article can't really be updated and this is one of the places where it's out of date.

**Reply by iguessthislldo** [reply]
I have put up a PR so the Developer's Guide mentions (#4092) this and have also put a note about the intro article not being kept up to date (#4093).

**Reply by MuhammedErkam** [reply]
I am using OpenDDS 3.16 and RTPS. As far as I see, 3.16 doesn't support UseXtypes. But I couldn't understand TypeObject. Is it topic?

**Reply by mitza-oci** [reply]
TypeObject is part of the DDS-XTypes specification.  Implementations without XTypes have no TypeObjects.

In order to answer your original question please provide a pull request that adds a test case to the repository.

**Reply by iguessthislldo** [reply]
[3.16 did have `UseXTypes`](https://github.com/OpenDDS/OpenDDS/blob/branch-DDS-3.16/dds/DCPS/RTPS/RtpsDiscovery.cpp#L636) and XTypes so this all should still apply. The only change I'm aware is in that version you'd have to pass `UseXtypes=0` instead of `UseXTypes=no` because it didn't support `no/minimal/complete`.

**Reply by mitza-oci** [reply]
Sorry for any potential confusion and thanks @iguessthislldo for clarifying.
My previous message was just to explain a bit more about what the phrase "does not provide a TypeObject" means in this context.
